[FLINK-6244] Emit timeouted Patterns as Side Output

dawidwys opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

dawidwys commented:
R: @kl0u 

@aljoscha I think it would be nice if you also had a look, as you spotted some problems with first approach to side-ouputs in CEP.

aljoscha commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java:
```diff
@@ -107,13 +113,54 @@
+	 * partial pattern sequence the provided {@link PatternTimeoutFunction} is called. The pattern
+	 * timeout function can produce exactly one resulting element.
+	 *
+	 * <p>You can get the stream of late data using
```
Maybe something like "You can get the stream of timed-out data resulting from the {@link PatternTimeoutFunction} using ..."

aljoscha commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java:
```diff
@@ -125,25 +172,66 @@
 	 * partial pattern sequence the provided {@link PatternTimeoutFunction} is called. The pattern
 	 * timeout function can produce exactly one resulting element.
 	 *
+	 * <p>You can get the stream of late data using
```
Same as above.

aljoscha commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java:
```diff
@@ -213,14 +321,97 @@
+	 * partial pattern sequence the provided {@link PatternFlatTimeoutFunction} is called. The pattern
+	 * timeout function can produce exactly one resulting element.
+	 *
+	 * <p>You can get the stream of late data using
```
Same as above.

aljoscha commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java:
```diff
@@ -213,14 +321,97 @@
+	 * partial pattern sequence the provided {@link PatternFlatTimeoutFunction} is called. The pattern
+	 * timeout function can produce exactly one resulting element.
+	 *
+	 * <p>You can get the stream of late data using
```
Same as above.

aljoscha commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/PatternStream.java:
```diff
@@ -163,14 +251,34 @@
-				patternStream.getExecutionEnvironment().clean(patternTimeoutFunction)
-			)
-		).returns(outTypeInfo);
+		return mainStream.connect(timeoutedStream).map(new CoMapFunction<R, L, Either<L, R>>() {
```
I think using an anonymous inner class here can be problematic because it will capture the `PatternStream` in it's closure, if I'm not mistaken.

dawidwys commented:
Thanks for the review @aljoscha . I've addressed all your comments. Also changed a section in docs to use the preferred version of handling timeouted results. I've rebased the PR, as it was far behind master and there were some conflicts.

aljoscha commented:
To me, this looks good now. Let's see if @kl0u want's to take another look, otherwise I would say it's good to merge.

dawidwys commented:
I will merge it then.

tedyu commented on docs/dev/libs/cep.md:
```diff
@@ -1279,63 +1279,75 @@ and `flatSelect` API calls allow a timeout handler to be specified. This timeout
+parameters
+
+ * `PatternTimeoutFunction`/`PatternFlatTimeoutFunction`
+ * [OutputTag]({{ site.baseurl }}/dev/stream/side_output.html) for the side output in which the timeouted matches will be returned
```
timeouted -> timed out

tedyu commented on flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/PatternStream.scala:
```diff
@@ -84,37 +79,54 @@ class PatternStream[T](jPatternStream: JPatternStream[T]) {
     *                               pattern sequence.
     * @tparam L Type of the resulting timeout event
     * @tparam R Type of the resulting event
+    * @deprecated Use the version that returns timeouted events as a side-output
```
timeouted -> timed out

tedyu commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java:
```diff
@@ -348,7 +353,18 @@ private void updateNFA(NFA<IN> nfa) throws IOException {
+
+		try {
+			processMatchedSequences(patterns.f0, timestamp);
+			processTimeoutedSequence(patterns.f1, timestamp);
```
processTimeoutedSequence -> processTimedoutSequence

tedyu commented on flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java:
```diff
@@ -348,7 +353,18 @@ private void updateNFA(NFA<IN> nfa) throws IOException {
+			nfa.process(event, timestamp);
+
+		try {
+			processMatchedSequences(patterns.f0, timestamp);
```
processMatchedSequences -> processMatchingSequences
